# Introduction to scRNA-seq Data

---------------------------------------------------
### Learning Objectives
- Familiarize yourself with different type of dissociation protocols
- Learn about different types of library preparation methods
- Acquaint yourself with the types of files generated by pre-processing pipelines (10X Genomics -- Cell Ranger)
---------------------------------------------------

# Cell Dissociation Methods
Th best way to achieve a good, clean scRNA-seq data set is to optimize your dissociation protocol for the tissues from which you are hoping to isolate cells. It is dually important to take into consideration the cell types you are hoping to isolate. For example, a hardier cell type in a tougher tissue may tolerate a harsh enzymatic digestion and result in a higher cell yield. There are many dissociation protocols available online at the 10X Genomics website linked [here](https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sample-prep). Today we will be discussing basic mechanical and enzymatic digestion. We will also review several types of reagents commonly used in enzymatic digestion.

- Mechanical Digestion: Mechanical digestion is the process of physically mincing, chopping, or pulverizing tissue to release cells from their organs or matrices.
- Enzymatic Digestion: Enzymatic digestion the process of freeing cells from their tissues using enzymes. Frequently, a combination of mechanical and enzymatic digestion is utilized to attain a single cell suspension. Below we have characterized a few reagents typically used for enzymatic digestion:
  - DNAse: DNAse I is used in tissue dissociation to degrade free DNA that exist in our suspension after dead cells are lysed. These pieces of DNA are often "sticky" and cause cells to clump together so getting rid of them is advantageous.
  - Collagenase: This class of enzymes specifically breaks down peptide bonds in collagen. A range of collagenases are commercially available and optimized for different tissues or cell types. It is important to consult the manufacturer's information to select the best collagenase for your dissociation.
  - Liberase: Liberase is a blend of enzymes that were designed to improve viability and functionality of isolated cells.
  - Dispase: Dispase cleaves fibronectin and collagen IV.
  - Accutase: This is a solution of proteolytic and collagenolytic enzymes that are useful for detatchment of cells from culture dishes and one another to a lesser extent.

** Note - Cells do not like being outside of a tissue environment or in enzymatic cocktails. Optimizing your dissociation method to reduce the time cells spend in solutions is beneficial. Take into careful consideration the type of tissue you are working with as well as the type of cells you wish to isolate.

- Filtering: After successfully dissociating your tissues it is vital to filter your suspension. This remove clumps of cells and large debris that can impede downstream library preparation.

# Library Preparation Methods
 - Combinatorial Barcoding - Individual cells are placed in their own barcoded wells. These barcodes are appended to the transcripts in the wells. This process is repeated two more times before the cells are pooled into 8 libraries. These cells are then lysed and receive an Illumina barcode. Transcripts that contain the same four barcode combination are from the same cell. See the below schematic to better understand his process. One example of this type of library preparation is Parse Bioscience. Here is a link to a video detailing this [process](https://www.youtube.com/watch?v=HVx4UBweNH4) and a link to the Parse Biosciences [website](https://www.parsebiosciences.com/).


 ![Parse](/images/Parse_Comb_Bar.png)

 - Droplet Based - A second type of library preparation is known as droplet based library preparation. Barcoded gel beads containing a read 1 primer, 10X barcode, unique molecular identifier (UMI), and a poly A sequence flow through a chamber and combine with a single cell in an oil bubble with enzymes. Thousands of reactions take place within these oil bubbles resulting in transcripts labeled with a cell barcode and a transcript barcode. These barcodes allow us to map back transcript to single cells. An example of droplet based library preparation is the 10X Genomics library preparation pipeline. A link to the 10X Genomics website detailing this process can be found [here](https://www.10xgenomics.com/)

 ![10X_Lib_Prep](/images/10X_Lib_Prep.jpeg)

 # Cell Ranger Pipeline
 ### Cell Ranger File Types

 - It is important to understand which files to work with from a Cell Ranger run. Cell Ranger provides a number of dynamic file types that allow you to interface with your data. A comprehensive description of each of these file types is provided in the table below:

 ![File_Des](/images/File_Des.png)

 - Web_summary.html: This is a useful tool for gaining insight into the quality of the single cell library you sequenced. Below is a screen shot example of a web_summary taken from the 10X Genomics website. The "Summary" tab contains several metrics for your data set including the estimated number of cells sequenced, mean reads per cell, and median genes per cell among others. If the estimated number of cells sequenced is very low, it is possible that you under loaded the chromium controller and if the estimated number of cells sequenced is very high is could mean that the controller was overloaded.

 ![web_summary](/images/websummary_stats.png)

  - Your web_summary.html will also display two t-SNE projections. One t-SNE is colored by UMI counts. This is the number of UMIs per cell. Cells with an unusually high UMI count could be doublets. When two cells stick together and are placed in the same oil bubble with one gel emulsification bead, the transcripts in both of those cells receive the same 10X cell-specific barcode. With two cells, this could mean double the UMIs, which will display in the UMI count t-SNE as red.

  - The second t-SNE plot is a cursory clustering of your sequenced cells. This clustering is subject to change after you begin downstream processing in Seurat or Monocle. For example, you will filter out potential doublets and negatives. There are options to filter out dead cells and subset on cells with lowly expressed genes. This t-SNE is accompanied by "Top Features by Cluster", or the genes that distinguish the cluster from one another.

 ![web_summary](/images/websummary_t-SNE.png)

 - Cloupe File: This file can be loaded into 10X Genomics Cell Loupe Browser for further exploration of your unprocessed data to assess it's quality. The cloupe file also provides information on differentially expressed genes in each of the Cell Ranger identified clusters. You may also wish to search genes or antibodies of interest, metrics you can not obtain from the web_summary alone.

 - Filtered_bc_matrix: These files are the files that you will read into R for downstream analysis. The filtered and raw matrix folders contain a matrix.tsv file with count data for each gene in a cell. The features.tsv file corresponds to row indices. The first two columns of this file are feature IDs and the third column is the type of feature (ex. "Gene Expression","Antibody Capture","CRISPR", etc.). The third and final file in the barcodes.tsv file, which contains a set of barcodes that correspond to column indices. You may choose to read in the folder containing these files or you may choose to read in the filtered_bc_matrix.h5 file, which contains the same information in one file.

 # Reading in scRNA-seq Data

 - Once you are largely satisfied that your data set is of good quality, meaning you have captured approximately the number of cells you intended to capture and your read depth is adequate, you can read your data into R using Seurat. [Surat](https://satijalab.org/seurat/) was developed by the Satija lab as an R toolkit for single cell genomics. The Seurat website is a valuable resource for tutorials, troubleshooting, and function information.

 Let's begin by reading in the example data set we will be working with in this workshop:

 ```{r}
## Loading Libraries
library(Seurat)
library(SeuratData)
library(ggplot2)
library(patchwork)
library(dplyr)
library(hdf5r)
 ```
```{r}
path = "/Users/noellekosarek/Desktop/GSE189460_RAW/"
files <- list.files(path=path)
object_list = c()

x = 1
for (i in files){
	if (x == 1){
		data <- Read10X_h5(paste0(path,i))
		first_object <- CreateSeuratObject(data, min.cells = 3, min.features = 200)
		x = x+1
	} else {
		data <- Read10X_h5(paste0(path,i))
		data_obj <- CreateSeuratObject(data, min.cells = 3, min.features = 200)
		object_list <- append(object_list, data_obj)
	}
  }

#Get sample names for each file read in
sample_names <- c()
for (i in files){
	print(i)
	sample_name <- gsub("_filtered_feature_bc_matrix.h5","",i)
	sample_names <- append(sample_names, sample_name)
}
print(sample_names)
object <- merge(first_object, object_list, add.cell.ids = sample_names)
```
